### PR TITLE
Add more cases to the `useless_conversion` lint

### PR DESCRIPTION
**Pull request by samueltardieu**
The new cases are the application of `Into::into` or `From::from` through the following functions with no effect:
- `Option::map()`
- `Result::map()` and `Result::map_err()`
- `ControlFlow::map_break()` and `ControlFlow::map_err()`
- `Iterator::map()`

changelog: [`useless_conversion`]: detect useless calls to `Into::into` and `From::from` application through `map*` methods

### CODE REVIEW

**Comment by samueltardieu** [comment]
The hits in lintcheck are true positives.

**Comment by samueltardieu** [comment]
Lookup code will be simplified when https://github.com/rust-lang/rust/pull/133686 is merged and synchronized with Clippy.

File: clippy_lints/src/useless_conversion.rs [hunk 1]
```diff
@@ -382,3 +384,58 @@ impl<'tcx> LateLintPass<'tcx> for UselessConversion {
...
+        if def_path
+            .iter()
+            .map(Symbol::as_str)
+            .zip(["core", "ops", "control_flow", "ControlFlow"])
+            .all(|(sym, s)| sym == s)
+        {
+            return true;
+        }
+    }
+    if is_trait_method(cx, expr, sym::Iterator) {
```
Comment (blyxyas) [inline clippy_lints/src/useless_conversion.rs:437]: Do we really want to lint for any methods of `Iterator`? What would the use case for this branch?

File: clippy_lints/src/useless_conversion.rs [hunk 2]
```diff
@@ -382,3 +384,58 @@ impl<'tcx> LateLintPass<'tcx> for UselessConversion {
...
+    if is_inherent_method_call(cx, expr)
+        && let Some(recv_ty_defid) = recv_ty.ty_adt_def().map(AdtDef::did)
+    {
+        if let Some(diag_name) = cx.tcx.get_diagnostic_name(recv_ty_defid)
+            && matches!(diag_name, sym::Option | sym::Result)
+        {
+            return true;
+        }
+
+        // Necessary for `core::ops::control_flow::ControlFlow` until a diagnostic item
```
[the remark this answers is not in the document]
Comment (samueltardieu) [reply clippy_lints/src/useless_conversion.rs:425]: Will do. My PR has been merged into `rust-lang` so we should get it at the next sync.

File: clippy_lints/src/useless_conversion.rs (hunk 1 above)
Comment (samueltardieu) [reply clippy_lints/src/useless_conversion.rs:437]: This only activates for the recognized methods (`map_*`), no other methods of `Iterator` will trigger that.

**Review by blyxyas — APPROVED** [review]
LGTM, thanks! :heart:
Could you squash that third commit into the second one?

**Comment by samueltardieu** [comment]
> Could you squash that third commit into the second one?

Done

**Pull request merged**